programs/drift_vaults/src/instructions/redeem_tokens.rs

```diff
@@ -0,0 +1,238 @@
+use crate::constraints::{
+    is_ata, is_authority_for_vault_depositor, is_mint_for_tokenized_depositor,
+    is_tokenized_depositor_for_vault, is_user_for_vault,
+};
+use crate::error::ErrorCode;
+use crate::state::traits::VaultDepositorBase;
+use crate::token_cpi::{BurnTokensCPI, TokenTransferCPI};
+use crate::{validate, AccountMapProvider};
+use crate::{TokenizedVaultDepositor, Vault, VaultDepositor, VaultProtocolProvider, WithdrawUnit};
+use anchor_lang::prelude::*;
+use anchor_spl::token::{burn, transfer, Burn, Mint, Token, TokenAccount, Transfer};
+use drift::instructions::optional_accounts::AccountMaps;
+use drift::math::safe_math::SafeMath;
+use drift::state::user::User;
+
+pub fn redeem_tokens<'info>(
+    ctx: Context<'_, '_, 'info, 'info, RedeemTokens<'info>>,
+    tokens_to_burn: u64,
+) -> Result<()> {
+    let clock = &Clock::get()?;
+
+    let mut vault = ctx.accounts.vault.load_mut()?;
+
+    validate!(!vault.in_liquidation(), ErrorCode::OngoingLiquidation)?;
+
+    let mut vault_depositor = ctx.accounts.vault_depositor.load_mut()?;
+    let mut tokenized_vault_depositor = ctx.accounts.tokenized_vault_depositor.load_mut()?;
+
+    // backwards compatible: if last rem acct does not deserialize into [`VaultProtocol`] then it's a legacy vault.
+    let mut vp = ctx.vault_protocol();
+    vault.validate_vault_protocol(&vp)?;
+    let vp = vp.as_mut().map(|vp| vp.load_mut()).transpose()?;
+
+    let manager_shares_before = vault.total_shares.safe_sub(vault.user_shares)?;
+    let total_shares_before = vault_depositor
+        .get_vault_shares()
+        .safe_add(tokenized_vault_depositor.get_vault_shares())?
+        .safe_add(manager_shares_before)?;
+
+    let user = ctx.accounts.drift_user.load()?;
+    let spot_market_index = vault.spot_market_index;
+    let AccountMaps {
+        perp_market_map,
+        spot_market_map,
+        mut oracle_map,
+    } = ctx.load_maps(clock.slot, Some(spot_market_index), vp.is_some())?;
+
+    let vault_equity =
+        vault.calculate_equity(&user, &perp_market_map, &spot_market_map, &mut oracle_map)?;
+
+    validate!(
+        !vault_depositor.last_withdraw_request.pending(),
+        ErrorCode::InvalidVaultDeposit,
+        "Cannot redeem tokens with a pending withdraw request"
+    )?;
+
+    let total_supply_before = ctx.accounts.mint.supply;
+
+    let (shares_transferred, shares_to_transfer) = match vp {
+        None => {
+            let (shares_to_transfer, _) = tokenized_vault_depositor.redeem_tokens(
```

Comment:
does the rebase happen here?

Reply:
yeah it's called in the tokenized_vault_depositor

https://github.com/drift-labs/drift-vaults/blob/928ab3c2d2f5b2e03aa3376acda2c5f1326db584/programs/drift_vaults/src/state/tokenized_vault_depositor.rs#L270